[KOGITO-9454]  Documenting java flow library (#585)
apache/incubator-kie-kogito-docs: serverlessworkflow/modules/ROOT/pages/getting-started/java-embedded-workflows.adoc

= Workflow embedded execution in Java 

This guide show cases how to execute a link:{spec_doc_url}[CNCF Serverless Workflow] definition using a standard java virtual machine and a small set of Maven dependencies. Therefore, it is assumed you are fluent both in Java and Maven. 
The workflow definition to be executed can be read from a .json or .yaml file or programatically defined using the {product_name} fluent API. 

[[embedded-file-quick-start]]
== Hello world (using existing definition file)

First step is to setup an empty Maven project with the following dependency. 

[source,xml]
----

> <dependency>

Author: @ricardozanini Yes, the renaming from kogito to sonataflow is far from completed

Author: The rename will be done with https://github.com/apache/incubator-kie-kogito-runtimes/issues/3447

  <groupId>org.kie.kogito</groupId>
  <artifactId>kogito-serverless-workflow-executor-core</artifactId>
  <version>RELEASE</version>
</dependency>
----

> Also, you might optionally add `simple logger for java` dependency to avoid using `system.out.println`

Reviewer: ```suggestion
Also, you might optionally add `simple logger for java` dependency to avoid using `System.out.println`
```

[source,xml]
----
 <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>slf4j-simple</artifactId>
      <version>1.7.36</version>
  </dependency>
----

> Lets assume you already have a workflow definition written in a JSON file in your project root. For example, link:{kogito_sw_examples_url}/serverless-workflow-hello-world/src/main/resources/hello.sw.json[Hello World] definition. In order to execute it, you need to write the following main java class (standard imports and java package declaration are intentionally skipped for brevity)

Reviewer: ```suggestion
Let's assume you already have a workflow definition written in a JSON file in your project root. For example, link:{kogito_sw_examples_url}/serverless-workflow-hello-world/src/main/resources/hello.sw.json[Hello World] definition. To execute it, you need to write the following main java class (standard imports and java package declaration are intentionally skipped for brevity)
```

[source,java]
----
      try (StaticWorkflowApplication application = StaticWorkflowApplication.create()) {
            Workflow workflow = workflow("HelloWorld"). <1>
            		start( <2>
            				inject( <3>
            						jsonObject().put("greeting", "Hello World").put("mantra","Serverless Workflow is awesome!"))) <4>
            		.end() <5>
            		.build(); <6>
            logger.info("Workflow execution result is {}",application.execute(workflow, Collections.emptyMap()).getWorkflowdata());  <7>
        }
----
<1> Creates a workflow which name is `HelloWorld`
<2> Indicate that you are going to specify the start state
<3> A Inject state is the start state
<4> Inject state accepts static json, therefore this line creates the JSON data
<5> End the workflow definition 
<6> Build the workflow definition
<7> Execute and print as in previous example

== Additional resources

> You can find additional and commented examples of fluent API usage (including jq expression evaluation and orchestation of rest services) link:https://github.com/fjtirado/java-flow/tree/main/java-flow-example/src/main/java/org/kie/kogito/serverless/workflow/examples[here]

Author: Wondering if we should keep this guide small and refer to further, already written examples or we prefer the "verbose" approach, with the risk of outdating

Reviewer 2: Either examples and docs can be outdated. But I guess having examples in our CI, makes them easier to spot and then update.

Author: I agree, let me move this classes to our examples repo and link them properly

Author: Created https://github.com/apache/incubator-kie-kogito-examples/pull/1895 for adding the examples

= Workflow embedded execution in Java 

> This guide show cases how to execute a link:{spec_doc_url}[CNCF Serverless Workflow] definition using a standard java virtual machine and a small set of Maven dependencies. Therefore, it is assumed you are fluent both in Java and Maven. 

Reviewer 3: ```suggestion
This guide uses a standard Java virtual machine and a small set of Maven dependencies to execute a link:{spec_doc_url}[CNCF Serverless Workflow] definition. Therefore, it is assumed you are fluent both in Java and Maven. 
```

> The workflow definition to be executed can be read from a .json or .yaml file or programatically defined using the {product_name} fluent API. 

Reviewer 3: ```suggestion
The workflow definition to be executed can be read from a `.json` or `.yaml` file or programmatically defined using the {product_name} fluent API. 
```

[[embedded-file-quick-start]]
== Hello world (using existing definition file)

> First step is to setup an empty Maven project with the following dependency. 

Reviewer 3: ```suggestion
The first step is to set up an empty Maven project with the following dependency:
```

[source,xml]
----
<dependency>
  <groupId>org.kie.kogito</groupId>
  <artifactId>kogito-serverless-workflow-executor-core</artifactId>

>   <version>RELEASE</version>

Reviewer 4: suggestion 1:  we could reference maven repository of this dependency  for better UX.

suggestion 2: use some variable from `antora.yaml` which is unified across the docs, `sonataflow_kie_apache_version=999.0.0-SNAPSHOT` ( for releases, we fix it to 10.0.0.Final ). Unfication can be done later.

</dependency>
----

> Also, you might optionally add a `simple logger for java` dependency to avoid using `System.out.println`

Reviewer 4: ```suggestion
This guide uses [slf4j dependency](https://mvnrepository.com/artifact/org.slf4j/slf4j-simple). Consider using any other logger for java to avoid using `System.out.println`
```

> Also, you might optionally add `simple logger for java` dependency to avoid using `System.out.println`

Reviewer 3: ```suggestion
Also, you might optionally add a `simple logger for java` dependency to avoid using `System.out.println`
```

[source,xml]
----
 <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>slf4j-simple</artifactId>

>       <version>1.7.36</version>

Reviewer 4: This version would need updating, since it is an example I think referencing maven repo of slf4j is enough. Wdyt?

  </dependency>
----

> Let's assume you already have a workflow definition written in a JSON file in your project root. For example, link:{kogito_sw_examples_url}/serverless-workflow-hello-world/src/main/resources/hello.sw.json[Hello World] definition. To execute it, you need to write the following main java class (standard imports and java package declaration are intentionally skipped for brevity)

Reviewer 3: ```suggestion
Let's assume you already have a workflow definition written in a JSON file in your root project. For example, link:{kogito_sw_examples_url}/serverless-workflow-hello-world/src/main/resources/hello.sw.json[Hello World] definition. To execute it, you must write the following main Java class (standard imports and Java package declaration are intentionally skipped for briefness)
```

    public static void main(String[] args) throws IOException {
        try (Reader reader = new FileReader("hello.sw.json"); <1>
                StaticWorkflowApplication application = StaticWorkflowApplication.create()) { <2>
            Workflow workflow = ServerlessWorkflowUtils.getWorkflow(reader, WorkflowFormat.JSON); <3>
            JsonNodeModel result = application.execute(workflow, Collections.emptyMap()); <4>
            logger.info("Workflow execution result is {}", result.getWorkflowdata()); <5>
        }
    }
}
----
<1> Reads the workflow file definition from the project root directory
<2> Creates a static workflow application object. It is done within the try block since the instance is `Closeable`. This is the reference that allow you to execute workflow definitions.
<3> Reads the Serverless Workflow Java SDK `Workflow` object from the file. 
<4> Execute the workflow, passing `Workflow` reference and no parameters (an empty Map). The result of the workflow execution: process instance id and workflow output model, can accessed using `result` variable. 
<5> Prints the workflow model in the configured standard output. 

> If you compile and execute this java class, you will see the following log in your configured standard output

Reviewer 3: ```suggestion
If you compile and execute this Java class, you will see the following log in your configured standard output:
```

----
Workflow execution result is {"greeting":"Hello World","mantra":"Serverless Workflow is awesome!"}
----

[[embedded-fluent-quick-start]]
== Hello world (using fluent API)

> Using the same Maven setup than in the previous section, you can programatically generate that workflow definition rather than loading it from a file definition by using link:{kogito_runtimes_url}/kogito-serverless-workflow/kogito-serverless-workflow-fluent/src/main/java/org/kie/kogito/serverless/workflow/fluent[fluent API]

Reviewer 3: ```suggestion
Using the same Maven setup as in the previous section, you can programmatically generate that workflow definition rather than loading it from a file definition by using the link:{kogito_runtimes_url}/kogito-serverless-workflow/kogito-serverless-workflow-fluent/src/main/java/org/kie/kogito/serverless/workflow/fluent[fluent API]
```

> Therefore, you can modify the previous example in a way that generates exactly the same output when it is executed, but rather creating a `FileReader` that reads the `Workflow` object, we create the `Workflow` object using java statements. The resulting modified main method is the following

Reviewer 3: ```suggestion
Therefore, you can modify the previous example to generate the same output when it is executed, but rather than creating a `FileReader` that reads the `Workflow` object, we create the `Workflow` object using Java statements. The resulting modified main method is the following
```
